Account for GOT VariantKind on osx-arm64

am11 opened the pull request:
Related to https://github.com/dotnet/runtime/issues/67232.

filipnavara commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64MachObjectWriter.cpp:
```diff
@@ -64,6 +64,10 @@ bool AArch64MachObjectWriter::getAArch64FixupKindMachOInfo(
     return true;
+  case FK_PCRel_4:
+    Log2Size = Log2_32(4);
+    RelocType = unsigned(MachO::ARM64_RELOC_POINTER_TO_GOT);
```
This doesn't look right. I'd expect it to follow similar logic to `FK_Data_4` and generate `ARM64_RELOC_UNSIGNED` by default [and `ARM64_RELOC_POINTER_TO_GOT` based on the symbol kind].

filipnavara replied:
(specifically, the 495e374f8fb370a0bf45304fbc835dbf78f2949d commit doesn't look right)

am11 replied:
These are experimental changes trying to move the needle a bit to get apps working. Here are the details for this one https://github.com/dotnet/runtime/issues/67232#issuecomment-1087985783.

If you want to play with it, you can checkout this PR branch in llvm-project, apply  https://github.com/dotnet/runtime/compare/main...am11:feature/nativeaot/osx-arm64 on runtime's `main`, and then use this script: https://github.com/dotnet/runtime/pull/68211#discussion_r854238425 to test the whole thing. Rinse and repeat.

Once you are there deep in lldb session, it would be easier to get the feel what kind of behavior we get by tweaking `FK_PCRel_4` case (spoiler alert; it ain't pretty :grinning:).

filipnavara replied:
I actually went through the issue and reproduced all the same failures as you did :)

am11 replied:
I might get back to it in a few days. Meanwhile, if you find the culprit, feel free to add more commits. I have added you as collaborator in my forks.

filipnavara commented on llvm/tools/objwriter/objwriter.cpp:
```diff
@@ -471,13 +505,19 @@ int ObjectWriter::EmitSymbolRef(const char *SymbolName,
   }
   case RelocType::IMAGE_REL_BASED_ARM64_PAGEBASE_REL21: {
+    if (ObjFileInfo->getObjectFileType() == ObjFileInfo->IsMachO) {
+      Kind = MCSymbolRefExpr::VK_GOTPAGE;
```
I assume this should be `VK_PAGE` and not `VK_GOTPAGE`. We likely want to generate `ARM64_RELOC_PAGE21` / `ARM64_RELOC_PAGEOFF12` and not a mix of `ARM64_RELOC_GOT_LOAD_PAGE21` 
 and `ARM64_RELOC_PAGEOFF12`.

am11 replied:
Can't remember why it was `VK_GOTPAGE`, but `VK_PAGE` is also working (current issues are unaffected by this change)
```suggestion
      Kind = MCSymbolRefExpr::VK_PAGE;
```

filipnavara replied:
The last rebase got this wrong again... aside from that it looks ok.

am11 replied:
Which particular test is failing with it? How is this "wrong"?

filipnavara replied:
It's not easily visible (at least not on small binaries) but it really should be VK_PAGE to match VK_PAGEOFF in the other relocation.

am11 commented:
@MichalStrehovsky this can be merged as well. We are emitting the correct kind, further changes are made in dotnet/runtime.

am11 commented:
Fixed an unrelated issue: mac legs were failing with
<img width="1133" alt="image" src="https://user-images.githubusercontent.com/3840695/189045615-f4cc67a6-0db3-4e2b-87c1-7e964721da3d.png">

I've updated the vmImage (macos-11 that matches with runtime). All green now.

MichalStrehovsky commented:
Cc @markples

filipnavara commented:
I still have concern about the VK_PAGE vs VK_GOTPAGE above (cannot mark the conversation as unresolved). The rest looks okay and I am reasonably confident that the changes produce working relocations.

am11 commented:
I have ran the tests with this set of changes and posted the smoke tests results here: https://github.com/dotnet/runtime/issues/67232#issuecomment-1236055626. Feel free to push further changes in a follow up PRs. Lets not drag it further.

filipnavara commented:
> I have ran the tests with this set of changes and posted the smoke tests results here: [dotnet/runtime#67232 (comment)](https://github.com/dotnet/runtime/issues/67232#issuecomment-1236055626). Feel free to push further changes in a follow up PRs. Lets not drag it further.

I ran the same set of tests with the VK_PAGE change (in fact, I used it for few days now). It makes no difference for that particular test cases. However, VK_GOTPAGE/VK_GOTPAGEOFF pair clearly does not work and produces incorrect relocations.

am11 commented:
Changed to VK_PAGE and resolved the conflict.

MichalStrehovsky reviewed (approved):
Thank you!
